src/qp/operators/SortMergeJoin.java

```diff
@@ -0,0 +1,226 @@
+package qp.operators;
+
+import qp.utils.*;
+
+import java.io.*;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Collections;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+public class SortMergeJoin extends Join {
+    static int filenum = 0;         // To get unique filenum for this operation
+    int batchsize;                  // Number of tuples per out batch
+    ArrayList<Integer> leftindex;   // Indices of the join attributes in left table
+    ArrayList<Integer> rightindex;  // Indices of the join attributes in right table
+    String rfname;                  // The file name where the right table is materialized
+    Batch outbatch;                 // Buffer page for output
+    Batch leftbatch;                // Buffer page for left input stream
+    Batch rightbatch;               // Buffer page for right input stream
+    ObjectInputStream in;           // File pointer to the right hand materialized file
+
+    int lcurs;                      // Cursor for left side buffer
+    int rcurs;                      // Cursor for right side buffer
+    boolean eosl;                   // Whether end of stream (left table) is reached
+    boolean eosr;                   // Whether end of stream (right table) is reached
+
+    public SortMergeJoin(Join jn) {
+        super(jn.getLeft(), jn.getRight(), jn.getConditionList(), jn.getOpType());
+        schema = jn.getSchema();
+        jointype = jn.getJoinType();
+        numBuff = jn.getNumBuff();
+    }
+
+    public ArrayList<String> generateSortedRuns(Operator base, ArrayList<Integer> id, String basename) {
+        int runnum = 0;
+        Batch inbatch = base.next();
+        int MAX_SIZE = numBuff*inbatch.size();
+        ArrayList<Tuple> runToBeStored = new ArrayList<>(MAX_SIZE);
+        ArrayList<String> runNames = new ArrayList<>();
+        while (inbatch != null) {
+             assert inbatch.size() <= inbatch.capacity();
```

Comment:
if i knew then i wouldn't have put this here LOL. I would have handled it. It's just a sanity check, to stop any errors from propagating downwards, so that I can isolate bugs.